### PR TITLE
wip

### PR DESCRIPTION
**Pull request by oli-clive-griffin**
## Description
Adds the option to train with gradient accumulation.

Terminology:
- `batch_size`: The effective / theoretical batch size
- `microbatch_size`: The size of each processed batch
- `gradient_accumulation_steps`: The number of steps

`batch size == microbatch_size * gradient_accumulation_steps`

## Motivation and Context
We want to train with larger batch sizes, particularly for LMs

## How Has This Been Tested?
local runs. Will do a larger sweep too and post below.

## Does this PR introduce a breaking change?
Nope

### CODE REVIEW

**Comment by oli-clive-griffin** [comment]
another report (in progress) after fixing the loss division: 
https://wandb.ai/goodfire/spd/reports/SPD-Run-Report---run_20250723_093429--VmlldzoxMzcwMTk5Nw==

(it's from commit `caca048`, which is equivalent to the current state. see: https://github.com/goodfire-ai/spd/compare/caca048..f686c0e)
